exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py

```diff
@@ -219,27 +243,448 @@ def export(
             if delay == self._MAX_RETRY_TIMEOUT:
                 return MetricExportResult.FAILURE
 
-            resp = self._export(serialized_data.SerializeToString())
-            # pylint: disable=no-else-return
-            if resp.ok:
-                return MetricExportResult.SUCCESS
-            elif self._retryable(resp):
-                _logger.warning(
-                    "Transient error %s encountered while exporting metric batch, retrying in %ss.",
-                    resp.reason,
-                    delay,
-                )
-                sleep(delay)
-                continue
+            if self._max_export_batch_size is None:
+                resp = self._export(serialized_data.SerializeToString())
+                # pylint: disable=no-else-return
+                if resp.ok:
+                    return MetricExportResult.SUCCESS
+                elif self._retryable(resp):
+                    _logger.warning(
+                        "Transient error %s encountered while exporting metric batch, retrying in %ss.",
+                        resp.reason,
+                        delay,
+                    )
+                    sleep(delay)
+                    continue
+                else:
+                    _logger.error(
+                        "Failed to export batch code: %s, reason: %s",
+                        resp.status_code,
+                        resp.text,
+                    )
+                    return MetricExportResult.FAILURE
+
+            # Else, attempt export in batches for this retry
             else:
-                _logger.error(
-                    "Failed to export batch code: %s, reason: %s",
-                    resp.status_code,
-                    resp.text,
-                )
-                return MetricExportResult.FAILURE
+                export_result = MetricExportResult.SUCCESS
+                for split_metrics_data in self._split_metrics_data(
+                    serialized_data
+                ):
+                    split_resp = self._export(
+                        split_metrics_data.SerializeToString()
+                    )
+
+                    if split_resp.ok:
+                        export_result = MetricExportResult.SUCCESS
+                    elif self._retryable(split_resp):
```

Comment:
Wouldn't we need to remove the success/failure batches from the entire list when we attempt a retry? For example, batch size of 1 and there are 2 metric datas. The first one returns retryable and the second one returns failure. When the loop returns to the outer loop, we don't want to retry the one that failed again right?

Reply:
Good call, no we don't want to retry a non-retryable failure!

I reorganized the unbatched-vs-batched and, for the latter, made `batch` the outer loop and `delay` the inner loop with breaks for clarity in [f1ef6c4](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/f1ef6c404faab53a8eb9685206e8e8010a0b5f77).

Reply:
Hi @lzchen any other changes to suggest?